Add revng-js-prerequisites component

mrjackv opened the pull request.

mrjackv committed:
Add revng-js-prerequisites component

This component will provide a way for revng components to install
javascript packages from npmjs

aleclearmind commented on .orchestra/config/components/revng_js_components.yml:
```diff
@@ -0,0 +1,34 @@
+  cd "$BUILD_DIR"
+  cp "$JS_DIR/package.json" "$JS_DIR/yarn.lock" .
+  echo '170f9c4509efc5065364e99ac32823bf8a37fcbf1a05dd7cf85a530e4173d851  package.json' | sha256sum -c -
+  PACKAGE_URLS=$(grep 'resolved' yarn.lock | awk '{print $2}' | tr -d '"')
```
Can we have a comment on how `yarn.lock` was generated and how to re-generate it?

mrjackv replied:
Added a `README.md` file besides the `yarn.lock` that explains how to regenerate it

aleclearmind commented on .orchestra/config/components/revng_js_components.yml:
```diff
@@ -0,0 +1,34 @@
+  while IFS= read -r URL; do
+    wget -q "$URL"
+  done <<< "$PACKAGE_URLS"
+  echo $PACKAGE_URLS | sed 's;.*/\(.*\)#\(\w*\);\2 \1;g' | sha1sum -c -
```
We kinda follow this convention in orchestra:
```suggestion
  echo $PACKAGE_URLS | sed 's|.*/\(.*\)#\(\w*\)|\2 \1|g' | sha1sum -c -
```

mrjackv replied:
Done

aleclearmind commented on .orchestra/config/components/revng_js_components.yml:
```diff
@@ -0,0 +1,34 @@
+#@overlay/match-child-defaults missing_ok=True
+---
+components:
+  revng-js-components: #@ single_build_component(**js_components_args())
```
```suggestion
  revng-js-prerequisites: #@ single_build_component(**revng_js_prerequisites_args())
```

mrjackv replied:
Renamed `components -> prerequisites` everywhere

aleclearmind commented on .orchestra/config/components/revng_js_components.yml:
```diff
@@ -0,0 +1,34 @@
+  JS_DIR="$ORCHESTRA_DOTDIR/support/js-components"
+  mkdir -p "$BUILD_DIR"
+  cd "$BUILD_DIR"
+  cp "$JS_DIR/package.json" "$JS_DIR/yarn.lock" .
```
Let's always use `cp -a`.

mrjackv replied:
Fixed

aleclearmind commented on .orchestra/config/components/revng_js_components.yml:
```diff
@@ -0,0 +1,34 @@
+  echo '170f9c4509efc5065364e99ac32823bf8a37fcbf1a05dd7cf85a530e4173d851  package.json' | sha256sum -c -
+  PACKAGE_URLS=$(grep 'resolved' yarn.lock | awk '{print $2}' | tr -d '"')
+  while IFS= read -r URL; do
+    wget -q "$URL"
```
Can you split `.orchestra/support/extract.sh` into `fetch.sh` and `extract.sh` (which uses `fetch.sh`)?
This way we cache the archives in `SRC_ARCHIVE_DIR`.

mrjackv replied:
Done

aleclearmind commented on .orchestra/config/components/revng_js_components.yml:
```diff
@@ -0,0 +1,34 @@
+#@yaml/text-templated-strings
+---
+#@ def js_components_args():
+configure: |
```
Please thoroughly comment what's going on

mrjackv replied:
Added comments, let me know if they are clear enough or need some more polishing

aleclearmind commented:
What's the `revng` side of this?
Remember that `revng` should not assume orchestra.

mrjackv committed:
Split the wget part of extract.sh into fetch.sh

This simplifies the cases where files need to be downloaded
(and cached) without the need to uncompress them

mrjackv commented:
> What's the revng side of this?
> Remember that revng should not assume orchestra.

the revng side (or in general any dependent component), will have to:
* Have a compatible `package.json` (in the sense that the dependencies have compatible versions to those installed on the cache folder)
* Whenever an install happens `yarn` as to be invokes with the `$YARNRC` file, like this `yarn install --use-yarnrc $YARNRC` so that the local cache is used and no network interaction happens

aleclearmind commented:
Please reopen targeting using `feature/typescript-generator` as source branch.
